### DISCUSSION TITLE
catching exceptions thrown by SQL INSERTs in CRUD Controllers

### DISCUSSION THREAD

**Discussion by prescriptionlifeline** [Laravel-Backpack/community-forum#1015] [Q&A (Help)]
I have a CRUD for a User model with SoftDeletes as a trait. email is a unique() column and when I'm trying to add a new user with an email of a soft deleted user I get a Illuminate\Database\UniqueConstraintViolationException with the following message:

```
SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'user@gmail.com' for key 'users.users_email_unique'
```
This totally makes sense but I'd like to be able to `catch` when this exception is being thrown and do an `update()` on the table when this happens but idk where I'd catch this exception.

Any ideas?

For that matter I think it'd also be better if, for non soft deleted users, it just gave a more useful error, saying something like "A user with this email already exists" but idk where I'd do that in my CRUD controller either.

**Comment by pxpm** [comment]
You should prevent that from happening by having validation in your Controller. 

In you create/update operations you should create something like a FormRequest with your rules for the fields. Most likely laravel `unique` validation would prevent this from happening and you can define your own messages: https://laravel.com/docs/11.x/validation#form-request-validation

See for example in our demo: https://github.com/Laravel-Backpack/demo/blob/main/app/Http/Requests/DummyRequest.php
 https://github.com/Laravel-Backpack/demo/blob/ea5765fa4b3ef651f7fb6fb2ffbbc94e7d0f6987/app/Http/Controllers/Admin/DummyCrudController.php#L63

If you need to so something custom when that validation error trigger you should probably write your custom rule, that will behave similar to Laravel `unique`, but instead of reporting the error, it could do something else (handling the error on your own) and reporting back "all fine here, continue ..." 👍 

Cheers

**Reply by prescriptionlifeline** [reply]
I created a new class that extends FormRequest, as in that example, and added this as a rule:

```php
'email' => ['required', 'string', 'lowercase', 'email', 'max:255', 'unique:users'],
```

The only problem is that now, when I try to make edits on an existing user (I'm using the same FormRequest for both create and update operations), I get this error:

```php
The email has already been taken.
```
If I'm adding a _new_ user that makes sense but giving this error on existing emails that are already unique to them isn't helpful.

In the [ProfileUpdateRequest.php that Laravel Breeze creates](https://github.com/laravel/breeze/blob/2.x/stubs/default/app/Http/Requests/ProfileUpdateRequest.php) there's this:

```php
'email' => ['required', 'string', 'lowercase', 'email', 'max:255', Rule::unique(User::class)->ignore($this->user()->id)],
```
I tried that but got an error when I tried to access `/admin/user/{id}/edit`:

```
Attempt to read property "id" on null
```

Any ideas?

**Reply by pxpm** [reply]
Yes, you are on the right track.

You can still use the same form request, but I would advise you to separate them into two, so it's easy to assert about which does what. 

Instead of `$this->user()` use `backpack_user()->id`.

Cheers

**Reply by prescriptionlifeline** [reply]
Now I'm getting this error:

```
SQLSTATE[42S02]: Base table or view not found: 1146 Table 'mywebsite.App\Http\Requests\User' doesn't exist

SELECT count(*) AS aggregate FROM `App\Http\Requests\USER` WHERE `email` = user
```
It's weird that USER is in all caps but, regardless, I did some digging and, in `config/backpack/base.php`, found this:

```
'user_model_fqn' => config('auth.providers.users.model'),
```
I mean idk if that has anything to do with the error but it' my best current guess. And I imagine that corresponds to `config\auth.php` which does this:

```
<?php

return [
...
    /*
    |--------------------------------------------------------------------------
    | User Providers
    |--------------------------------------------------------------------------
    |
    | All authentication drivers have a user provider. This defines how the
    | users are actually retrieved out of your database or other storage
    | mechanisms used by this application to persist your user's data.
    |
    | If you have multiple user tables or models you may configure multiple
    | sources which represent each model / table. These sources may then
    | be assigned to any extra authentication guards you have defined.
    |
    | Supported: "database", "eloquent"
    |
    */

    'providers' => [
        'users' => [
            'driver' => 'eloquent',
            'model' => App\Models\User::class,
        ],
...
```
So from my vantage point everything looks good.

Any ideas?

**Reply by pxpm** [reply]
No, that has nothing to do with your error. 

And I am pretty sure it's something in your code. I never saw that syntax for "unique" rule, AFAIK in Laravel docs https://laravel.com/docs/11.x/validation#rule-unique, It should be something like: `Rule::unique('users')->ignore(backpack_user()->id)`

Cheers

**Reply by prescriptionlifeline** [reply]
`backpack_user()->id` doesn't work - it returns the id of the _currently logged in user_ - not of the user being edited.

**Comment by jcastroa87** [comment]
Im glad you solve the problem. I will close the issue, but please feel free to re-open or create a new one if needed.

Cheers.
